Common.Compat: Extract reusable compat utilities and rename to sdk

kaxil opened the pull request:
Extract compatibility import logic into reusable _compat_utils module and rename lazy_compat to sdk for better clarity. The new architecture makes it easier to maintain consistency across all compatibility modules without introducing cycles due to which mypy was failing in https://github.com/apache/airflow/pull/56867

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

eladkal commented on providers/common/compat/src/airflow/providers/common/compat/sdk.py:
We'll need to set next version in `provider.yaml` and then bump this as min version for all depended providers otherwise users might install newer version of depended provider that imports this file but with older version of `common.compat`

potiuk replied:
As discussed elsewhere - there are two options:

1) do the bump now everywhere (but this makes release preparation harder as you know - because you have to classify the changes more manually - that was the original design
2) leave # TODO in the dependent providers to bump common-compat when release manager prepares release


I think option 2) is better - to leave the "bumping" to release manager - but it's up to you as well and we need to finally settle on a single approach I think.

kaxil replied:
Yeah I am fine with either -- let me know what you prefer

kaxil committed:
Common.Compat: Extract reusable compat utilities and rename to sdk

Extract compatibility import logic into reusable _compat_utils module and rename lazy_compat to sdk for better clarity. The new architecture makes it easier to maintain consistency across all compatibility modules without introducing cycles due to which mypy was failing in https://github.com/apache/airflow/pull/56867

jscheffl reviewed (approved):
`sdk`as term is much better! Thanks!
